include/smp/lock.h

```diff
@@ -71,14 +71,6 @@ static inline void *sel4_atomic_exchange(void *ptr, bool_t
     while (!try_arch_atomic_exchange_rlx(&big_kernel_lock.head,
                                          (void *) big_kernel_lock.node_owners[cpu].node,
                                          (void **) &prev)) {
-        if (clh_is_ipi_pending(cpu)) {
-            /* we only handle irq_remote_call_ipi here as other type of IPIs
-             * are async and could be delayed. 'handleIPI' may not return
-             * based on value of the 'irqPath'. */
-            handleIPI(CORE_IRQ_TO_IRQT(cpu, irq_remote_call_ipi), irqPath);
```

Comment:
It's actually incorrect to call `handleIPI` here the value of `big_kernel_lock.node_owners[cpu].next` hasn't been correctly initialized until `sel4_atomic_exchange` returns and so `ipiStallCoreCallback` called by `handleIPI` will poll the incorrect lock address.

Reply:
I agree that it is incorrect, but do we have any data on why we were trying to do `handleIPI` in here? Were we trying to assure progress of something that otherwise wouldn't happen?

Reply:
The commit history says something that it can take up to seconds until the atomic exchange succeeds. It does not say which platform, maybe one that was quite bad at SMP support. Modern platform should be better, ARMv8.1 even brings the swap instruction back.

Reply:
@axel-h do you have a link for that message? The commit that introduced the handleIPI call doesn't have a body: https://github.com/seL4/seL4/commit/3d2ae69f9cb184c34b6b56d1365e57693e832db4#

Reply:
I was wondering about this code and eventually I found this discussion that has some background: https://sel4.discourse.group/t/why-does-the-clh-lock-not-use-gcc-for-arm-atomics/103, then I created https://github.com/seL4/seL4/issues/856

Reply:
This those long delays may be fixed by commits [Repair barriers in sel4_atomic_exchange](https://github.com/seL4/seL4/commit/8ba22dcdc5fa6) and [Repair barriers in clh_lock_acquire](https://github.com/seL4/seL4/commit/fe1278326f664b).